New document design system

minhngocd opened the pull request:
## What

Transition the “Add new document” page template to the Design System. [Trello](https://trello.com/c/764XzLue/3419-design-system-add-new-document)

## Approach

This PR is the first step of a multi-step process. 
The first step covered here is to replace the existing fields with design system components and/or publishing gem components. 

1. Feature Flag was added in [the previous PR](https://github.com/alphagov/specialist-publisher/pull/3084) via user permissions to hide new design system work in progress while we transition pages. Layout and view renders are toggled within controllers. 
2. For any new view that is touched to change to the design system, a duplicate `*_legacy` version is created and moved to a legacy folder
3. Any associated tests are duplicated and tests for the new design systems are created as `*_design_system_spec` and log in via the helper methods such as `log_in_as_design_system_editor`. legacy tests are preserved for regression testing
4. We are only dealing with styling of new document page fields for this PR and creating a new page happy path. 
5. The following has been descoped into other tickets : 
  * Preview button working and looking correct
  * Retaining data previously entered in fields
  * validation & error messages
  * Add another full functionality

## Why

This is required as part of ongoing improvements to adopt Design System standards across the publishing apps. This will also enable us to implement new scheduling functionality for Specialist Publisher that is consistent with Whitehall

## Screenshots

With no `preview_design_system` permissions
<img width="600" alt="image" src="https://github.com/user-attachments/assets/9308a7fe-b13f-4553-89dd-10d5c5c8387f" />

With `preview_design_system` permissions
<img width="600" alt="image" src="https://github.com/user-attachments/assets/1840df36-c6b8-4c1f-9f92-375c524f51ae" />

Add another component
<img width="851" alt="image" src="https://github.com/user-attachments/assets/5abd12a1-0a47-4892-a1b3-150c63240c13" />

Add another component restyled for Request new finder form
<img width="360" alt="image" src="https://github.com/user-attachments/assets/b4512b32-abed-4c13-aafb-3d249b7c5109" />

New replaced javascript length counter for summary
<img width="791" alt="image" src="https://github.com/user-attachments/assets/c605e29f-f9d6-4b01-8ef2-fa05d9d8652a" />



⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

matthillco committed:
Update design system layout and title

Change styling for design system layout and associated attachment links on the sidebar

matthillco committed:
Replace existing breadcrumbs with gem breadcrumb component

matthillco committed:
Update design system attachment_links view

Update title to use design system heading component

matthillco committed:
Replace Title input field for design system

Replace document title input field. Edit page is still on the legacy layout so should be using legacy form fields partial.

matthillco committed:
Replace Summary text area

Using the word count component and removing usage of the length counter js. The length counter js is only included in the legacy_application.js so should be removed as part of transition cleanup.

matthillco committed:
Replace Body textarea

matthillco committed:
Replace Language select component

matthillco committed:
Replace the single select component

Being the first element of specialist document facet fields being replaced, this commit also includes moving all legacy document views into a legacy metadata fields folder which in turn uses legacy form fields and specialist document fields.

The new design system view currently does not support facet validation for the new input fields so we are commenting out the tests that do facet validation. This will be added back in once we implement field validation and form validation, descoped as part of another ticket.

matthillco committed:
Replace generic textareas

matthillco committed:
Make format_name available to all fields

matthillco committed:
Replace the Save button

matthillco committed:
Replace date input

This date input field is ported over from Whitehall.
https://github.com/alphagov/whitehall/blob/main/app/models/concerns/date_validation.rb
https://github.com/alphagov/whitehall/blob/main/app/views/components/_datetime_fields.html.erb
This is ported along with its tests. However validation is ported but currently not in use yet as the validation piece has been descoped to another ticket.

matthillco committed:
Replace multiple select with Add Another

Add another component has been added to the new document form. The component has a lot of spacing around it which means it doesn't sit well within the component. We are removing the additional styling we've added to the component when we previously added them to the change request form. This will put the styling for the add another component consistent with the Whitehall fatality notice Add Another component.

They were also hooked into a .js class and were overriding a gem component, both of which are discouraged practices. Gem components are built aorund a concept of 'component isolation' which precludes this sort of ad-hoc CSS override. Any styling of a publishing gem component should be done in the gem, not within the apps.

matthillco committed:
Ported govspeak help component from Whitehall

This also removes the dependency on `govuk_admin_template` on the attachments page.
We now have our own govspeak help component, following the same styling as Whitehall.

minhngocd committed:
Switch legacy and new design system view naming

This is to make it easier to cherry-pick commits since the commits that transitions the view to the new design system is done on a file named 'new'. There would be a lot of effort to move the changes between files if we kept the convention as `new` and `new_design_system`. Switching to `new_legacy` and `new`

ChrisBAshton commented on app/views/documents/new_legacy.html.erb:
In the interests of "optimising code for deletion", it might be cleaner to encapsulate these legacy templates in a subdirectory, i.e. `app/views/documents/legacy/new.html.erb`

ChrisBAshton commented on app/views/documents/new.html.erb:
```diff
@@ -1,8 +1,19 @@
+      },
+      {
+        title: "New document",
+        url: documents_path(current_format.admin_slug)
```
Shouldn't these `url` values be diffeent?

minhngocd replied:
I can pick this up as Matt's out - I think the "New Document" link shouldn't have a URL if we want to mimic the old functionality?

ChrisBAshton commented on app/views/shared/_form_fields.html.erb:
```diff
@@ -7,9 +7,21 @@
+    count_message_threshold:,
+    count_message_selector:,
+  },
+} do %>
```
Is the character count JS functionality natively supported? I don't see an option for it on https://components.publishing.service.gov.uk/component-guide/textarea

Would be good to make clear in this commit ("Replace Summary text area") whether we're intending to retain the JS counter and what 'legacy' code from Specialist Publisher we're having to retain to support that.

minhngocd replied:
How do we make this call? Should I rope in Patrick?

ChrisBAshton replied:
I'd probably raise it with Andrew at standup tomorrow. If it's not a supported feature of govuk_publishing_components then it feels like a non-standard feature to me and we're best off dropping it - but it's a PM call IMO.

minhngocd replied:
Need to remove the old JS and replace it with the new counter component

ChrisBAshton commented on spec/features/creating_a_new_document_design_system_spec.rb:
```diff
@@ -75,14 +75,14 @@
-          expect(page).to have_css("div.field_with_errors span.elements-error-message", text: /#{facet['key'].humanize} can't be blank|#{facet['name']} can't be blank/)
-        end
-      end
+      # schema.facets.each do |facet|
```
Was this commenting out intentional? If so, good to make that clear in the commit message ("Replace the single select component" commit)

minhngocd replied:
Yes it is intentional as we've descoped validation as part of a separate card. I'll amend the commit message

ChrisBAshton commented on app/models/concerns/date_validation.rb:
Is this code - or any of the code in `app/views/shared/_datetime_fields.html.erb` - based on existing code elsewhere in alphagov? I suspect it is, as it's a lot of code to be added in one commit, and it looks pretty mature. If it is based on code elsewhere, I'd like to that existing code, in the "Replace date input" commit.

minhngocd replied:
It is, it's been ported over from Whitehall - i'll amend the commit message

minhngocd replied:
Looks like we're not using the validation part of it for now... I've amended that information to the commit message as well

ChrisBAshton commented on app/assets/stylesheets/application.scss:
```diff
@@ -28,6 +28,5 @@ $govuk-page-width: 1140px;
 @import 'govuk_publishing_components/components/summary-card';
 @import 'govuk_publishing_components/components/textarea';
 
-@import "./components/add-another";
```
I'd like to see a reason/"why" being added to the "Replace multiple select with Add Another" commit message, as it lacks detail at the moment. I'm also confused as to why this line of code is being removed 🤔

minhngocd replied:
Matt found that these style overrides were causing the add another components to look out of place on the form with some different margins and padding. Removing them makes them more consistent with how it is currently looking on Whitehall (used in fatality notices). We think these were put in when they were implemented for the admin change request pages. I'll also include some screenshots of those pages

lauraghiorghisor-tw replied:
I don't think there is a way around it, but I do find that the Add another component makes the form very busy and difficult to follow. I don't know if it's customary that this component lives at the bottom of the page, or if we can think up some way to delimit one select from another. If we have a document with three rounds of multiselect, the form will be an utter mess (and we probably do...).

Something to maybe consider when we finish migrating. Some UX/wrap-up card or a new card to tackle this.
I'd actually be in favour of using the other type of multi-select (that we have on the orgs in whitehall.)

minhngocd replied:
There's another 2 cards in relation to "Add another" component in Ready to pick up - might be a good place for us to get together and discuss the component specifically. I would descope it from this PR - will pop in some time in the calendar for all of us (Chris, Matt, myself + Laura if you're interested) to figure out what we want to do with it

matthillco replied:
Just chiming in here: As Minno said, we removed the styles as they looked wrong compared to Whitehall. They were also hooked into a `.js` class and were overriding a gem component, both of which are discouraged practices. Gem components are built aorund a concept of 'component isolation' which precludes this sort of ad-hoc CSS override. Any styling of a publishing gem component should be done in the gem, not within the apps.

Having said all that, I agree that the Add Another component has some strange UX and I'm not personally a fan. It would be good if we could get some UR on how users are interacting with it and where any pain points are. On a personal level I'd prefer to replace it with the Multiple variant of `select-with-search`.

ChrisBAshton commented on app/views/shared/_govspeak_help.html.erb:
I'd rename this commit from "Replace govspeak help component" to something like "Copy govspeak_help component into Specialist Publisher" so it's clearer we're not just swapping one component call for another pre-existing component.

ChrisBAshton reviewed (commented):
Nice one for persevering with this - it's a big job! ⭐

The PR would benefit from some before-and-after screenshots 🙏  Otherwise come comments below (mostly about maintaining a good audit trail in our commit messages)

lauraghiorghisor-tw commented on app/views/shared/_specialist_document_form.html.erb:
```diff
@@ -8,9 +8,14 @@
-    <%= render layout: "shared/form_group", locals: { f: f, field: field, label: field_config["name"] } do %>
-      <%= f.text_area field, class: 'form-control' %>
-    <% end %>
+    <%= render "govuk_publishing_components/components/textarea", {
```
For these generic ones, can we set the row limit to lower (3 or so)? I think they were a tad smaller on the original views.

minhngocd replied:
I'll rope in Matt tomorrow

minhngocd committed:
Move all legacy views into a legacy folder

In order to keep more legacy files together for ease of deletion but also keeping them as close as possible to the current usages, I've opted for creating legacy folders within sub view folders as opposed to a top-level legacy folder.

minhngocd commented on app/views/documents/new.html.erb:
```diff
@@ -16,10 +16,6 @@
 
-<% content_for :document_ready do -%>
-  GOVUK.formChangeProtection.init($('.new_document'), 'You have unsaved changes that will be lost if you leave this page.');
-<% end -%>
```
We probably also need to have a chat around the removal of the javascript popup when navigating away from unsaved changes too. It's causing issues with the chrome browser tests, and Whitehall does not have this functionality.

minhngocd replied:
Decision to remove and see whether or not people ask for it back. Make it consistent with Whitehall, doesn't cause issues with feature specs

minhngocd committed:
Design system: Remove Javascript popup for unsaved alerts when navigating away from the page

form_change_protection.js belongs to legacy application js and we don't plan on porting it over to the new design system for a number of reasons:
- consistency with Whitehall - Whitehall does not have this feature
- javascript pop ups may not be accessible
- javascript pop ups are causing issues for the feature specs trying to close the pop up modal
- we remove it and see if users need the change. if there is request, we can always bring it back.

minhngocd committed:
Design system: tweak layout

Make it more consistent with the way Whitehall has it.

lauraghiorghisor-tw reviewed (approved):
LGTM 

Sanity checked commits and changes ✅
